flink-runtime/src/main/java/org/apache/flink/streaming/api/lineage/DefaultLineageGraph.java

```diff
@@ -64,6 +65,16 @@ public List<LineageEdge> relations() {
         return Collections.unmodifiableList(lineageEdges);
     }
 
+    void addSources(List<SourceLineageVertex> partialSources) {
+        this.deduplicatedSources.addAll(partialSources);
+        this.sources = new ArrayList<>(deduplicatedSources);
```

Comment:
shouldn't this be `addAll` ?

Reply:
The sources is reconstructed from the deduplicated source to guarantee new duplications.